rollout/controller.go

```diff
@@ -267,15 +288,44 @@ func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
 	}
 	log.Info("Started Rollout workers")
 
-	gvk := schema.ParseGroupResource("virtualservices.networking.istio.io").WithVersion(c.defaultIstioVersion)
-	go controllerutil.WatchResourceWithExponentialBackoff(stopCh, c.dynamicclientset, c.namespace, gvk, c.rolloutWorkqueue, c.rolloutsIndexer, virtualServiceIndexName)
+	// Auto-detect whether user adds Istio Virtual Service to cluster
+	// If it is added, then run the istioVirtualServiceInformer so that we can use istioVirtualServiceLister
+	go c.runVirtualServiceInformer(stopCh)
 
 	<-stopCh
 	log.Info("Shutting down workers")
 
 	return nil
 }
 
+func (c *Controller) runVirtualServiceInformer(stopCh <-chan struct{}) {
+	for !c.istioVirtualServiceInformer.HasSynced() {
+		// Should only execute if Istio is not installed on cluster
+		if !istioutil.DoesIstioExist(c.dynamicclientset, c.defaultIstioVersion) {
+			time.Sleep(10 * time.Minute)
+		} else {
+			c.istioVirtualServiceInformer.Run(stopCh)
+			cache.WaitForCacheSync(stopCh, c.istioVirtualServiceInformer.HasSynced)
+		}
+	}
+}
+
+func (c *Controller) processNextEnqueuedObj(obj interface{}) {
```

Comment:
The method is too generically named. Controller.processNextEnqueuedObj is only dealing with istio objects, and should be named as such.